### PR TITLE
#7: Add circe support for enums used as keys of a Map (closes #7)

### LINKED ISSUE #7: Add circe support for enums used as keys of a Map
## requirements
Enumero should provide support for encoding/decoding of a `Map` having a `CaseEnum` as key.

## specs
provide implicit `KeyEncoder[E <: CaseEnum]` and `KeyDecoder[E <: CaseEnum]`

## misc
{optional: other useful info}

### PR DESCRIPTION
**Pull request by fes300**
Closes #7

### tests performed
screenshot 2019-01-16 at 18 09 37

### CODE REVIEW

**Comment by fes300** [comment]
sorry there is a little noise due to `scalafmt` formatting 😓

File: build.sbt [hunk 1]
```diff
@@ -56,7 +58,9 @@ lazy val circeSupport = project
   .settings(
     name := "enumero-circe-support",
     libraryDependencies ++= Seq(
-      "io.circe" %% "circe-core" % "0.9.0"
+      "io.circe" %% "circe-core" % "0.9.0",
+      "io.circe" %% "circe-parser" % "0.9.0",
```
Comment (gabro) [inline build.sbt]: This should be a test dependency

**Review by gabro — CHANGES_REQUESTED** [review]
Tip for the next time, don't re-format everything (or if you do, do it in a different commit), otherwise it's really hard to spot the actual diff.

Also, it seems like you're using the default `.scalafmt.conf`, which produces horribly aligned arguments when defining methods. I'm 👍 for introducing scalafmt, but can you please copy the configuration from our apiseed?

File: build.sbt (hunk 1 above)
Comment (fes300) [reply build.sbt]: how are you so fast??

Comment (gabro) [reply build.sbt]: I let other people work.

File: circeSupport/src/main/scala/io.buildo.enumero/CirceSupport.scala [hunk 2]
```diff
@@ -15,21 +16,37 @@ trait CirceSupport {
 
   implicit def decodeCaseEnum[E <: CaseEnum](implicit s: CaseEnumSerialization[E]): Decoder[E] =
     Decoder.decodeString.emap { str =>
-      s.caseFromString(str).map(Right(_)).getOrElse(Left(deserializationError(str, s.name)))
+      s.caseFromString(str)
+        .map(Right(_))
+        .getOrElse(Left(deserializationError(str, s.name)))
     }
 
+  implicit final def decodeKeyCaseEnum[E <: CaseEnum](
```
Comment (fes300) [inline circeSupport/src/main/scala/io.buildo.enumero/CirceSupport.scala:23]: lines 24 to 33 + minor changes in `build.sbt` + new tests in `CirceSupportSpec.scala` are the only things that you have to review @bytecodeguru  (other changes are just `scalafmt` formatting)

File: circeSupport/src/main/scala/io.buildo.enumero/CirceSupport.scala [hunk 3]
```diff
@@ -1,5 +1,6 @@
 package io.buildo.enumero
 
+import io.circe.{KeyDecoder, KeyEncoder}
 import io.circe._
```
Comment (bytecodeguru) [inline circeSupport/src/main/scala/io.buildo.enumero/CirceSupport.scala]: This import seems unnecessary, `{KeyEncoderm KeyDecoder}` should be already in scope thanks to
```scala
import io.circe._
```

File: circeSupport/src/test/scala/CirceSupportSpec.scala [hunk 4]
```diff
@@ -0,0 +1,48 @@
+import io.buildo.enumero._
+import io.buildo.enumero.circe._
+import io.circe._, io.circe.syntax._, io.circe.parser._
```
Comment (bytecodeguru) [inline circeSupport/src/test/scala/CirceSupportSpec.scala]: Please use different rows for this multiple package import.

File: circeSupport/src/test/scala/CirceSupportSpec.scala [hunk 5]
```diff
@@ -0,0 +1,48 @@
+import io.buildo.enumero._
+import io.buildo.enumero.circe._
```
Comment (bytecodeguru) [inline circeSupport/src/test/scala/CirceSupportSpec.scala:2]: Why do we need to `import io.buildo.enumero.circe._` ?

File: circeSupport/src/test/scala/CirceSupportSpec.scala [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    case object Earth extends Planet
+  }
+
+  val planetMap = Map[Planet, Int](
+    Planet.Mercury -> 12,
+    Planet.Venus -> 812763,
+    Planet.Earth -> 0
+  )
+
+  val json = planetMap.asJson
```
Comment (bytecodeguru) [inline circeSupport/src/test/scala/CirceSupportSpec.scala]: I'd move this into the test, serialisation should break if `CirceSupport` is not working as intended.

**Review by bytecodeguru — CHANGES_REQUESTED** [review]
Well done 👍 
I've left a couple of suggestions.

File: circeSupport/src/test/scala/CirceSupportSpec.scala (hunk 5 above)
Comment (fes300) [reply circeSupport/src/test/scala/CirceSupportSpec.scala:2]: we actually only use `DecodingFailure` and `Json`, I'll make it explicit 👍

**Review by bytecodeguru — APPROVED** [review]
💯

**Pull request merged**